File: src/models/project.py
from dataclasses import dataclass, field
from typing import Optional
import json
import uuid


@dataclass
class Point:
    x: float
    y: float
    index: int
    label: Optional[str] = None
    category: Optional[str] = None


@dataclass
class Measurement:
    id: str
    type: str                            # "line" | "polyline" | "polygon"
    label: str                           # fragment name, e.g. "Frag-01"
    points: list[tuple[float, float]]    # image-space pixel coords [(x, y), ...]
    value: float                         # length (unit) or area (unit²)
    unit: str                            # "cm" or "m"
    species: str = ""                    # genus / species name
    auto_width: float = 0.0             # oriented/bbox width in real units
    auto_height: float = 0.0            # oriented/bbox height in real units
    area: float = 0.0                   # polygon area in real units²
    perimeter_len: float = 0.0          # polygon perimeter in real units
    angle: float = 0.0                  # tilt of the height axis, degrees from horizontal

    @staticmethod
    def new(mtype: str, label: str,
            points: list[tuple[float, float]],
            value: float, unit: str,
            species: str = "",
            auto_width: float = 0.0,
            auto_height: float = 0.0,
            area: float = 0.0,
            perimeter_len: float = 0.0,
            angle: float = 0.0) -> "Measurement":
        return Measurement(id=str(uuid.uuid4()), type=mtype, label=label,
                           points=points, value=value, unit=unit,
                           species=species,
                           auto_width=auto_width, auto_height=auto_height,
                           area=area, perimeter_len=perimeter_len, angle=angle)


@dataclass
class ImageAnnotation:
    image_path: str
    points: list[Point] = field(default_factory=list)
    image_width: int = 0
    image_height: int = 0
    scale_factor: float = 1.0   # pixels per scale_unit; 1.0 = not calibrated
    scale_unit: str = "cm"      # "cm" or "m"
    measurements: list[Measurement] = field(default_factory=list)

    def labeled_count(self) -> int:
        return sum(1 for p in self.points if p.label is not None)

    def is_complete(self) -> bool:
        return bool(self.points) and self.labeled_count() == len(self.points)

    def coverage_stats(self) -> dict:
        labeled = [p for p in self.points if p.label is not None]
        if not labeled:
            return {}
        total = len(labeled)
        stats: dict[str, int] = {}
        for p in labeled:
            key = p.label or ""
            stats[key] = stats.get(key, 0) + 1
        return {k: round(v / total * 100, 2) for k, v in stats.items()}


@dataclass
class Station:
    name: str
    depth_m: Optional[float] = None
    date: Optional[str] = None      # ISO-8601: "2024-03-15"
    gps_lat: Optional[float] = None
    gps_lon: Optional[float] = None
    notes: str = ""
    annotations: list[ImageAnnotation] = field(default_factory=list)

    def total_points(self) -> int:
        return sum(len(a.points) for a in self.annotations)

    def labeled_points(self) -> int:
        return sum(a.labeled_count() for a in self.annotations)

    def is_complete(self) -> bool:
        return bool(self.annotations) and all(
            a.is_complete() for a in self.annotations
        )


@dataclass
class Project:
    name: str
    point_count: int = 10
    point_distribution: str = "random"  # random | stratified | uniform
    border_exclusion: int = 0           # uniform pixel border to exclude
    border_rect: list | None = None     # [x_min, y_min, x_max, y_max] if set by click
    border_polygon: list | None = None  # [[x, y], ...] if set by polygon drawing
    coral_codes: dict = field(default_factory=dict)
    coral_groups: list = field(default_factory=list)  # [{"name": str, "codes": [str]}]
    species_list: list[str] = field(default_factory=list)  # known species/genus names
    stations: list[Station] = field(default_factory=list)
    save_path: Optional[str] = None

    @property
    def annotations(self) -> list[ImageAnnotation]:
        """Flat view across all stations — for statistics and export."""
        return [a for s in self.stations for a in s.annotations]

    def save(self, path: str) -> None:
        data = {
            "name": self.name,
            "point_count": self.point_count,
            "point_distribution": self.point_distribution,
            "border_exclusion": self.border_exclusion,
            "border_rect": self.border_rect,
            "border_polygon": self.border_polygon,
            "coral_codes": self.coral_codes,
            "coral_groups": self.coral_groups,
            "species_list": self.species_list,
            "stations": [
                {
                    "name": s.name,
                    "depth_m": s.depth_m,
                    "date": s.date,
                    "gps_lat": s.gps_lat,
                    "gps_lon": s.gps_lon,
                    "notes": s.notes,
                    "annotations": [
                        {
                            "image_path": a.image_path,
                            "image_width": a.image_width,
                            "image_height": a.image_height,
                            "scale_factor": a.scale_factor,
                            "scale_unit": a.scale_unit,
                            "points": [
                                {"x": p.x, "y": p.y, "index": p.index,
                                 "label": p.label, "category": p.category}
                                for p in a.points
                            ],
                            "measurements": [
                                {"id": m.id, "type": m.type, "label": m.label,
                                 "points": m.points, "value": m.value, "unit": m.unit,
                                 "species": m.species,
                                 "auto_width": m.auto_width, "auto_height": m.auto_height,
                                 "area": m.area, "perimeter_len": m.perimeter_len,
                                 "angle": m.angle}
                                for m in a.measurements
                            ],
                        }
                        for a in s.annotations
                    ],
                }
                for s in self.stations
            ],
        }
        with open(path, "w", encoding="utf-8") as f:
            json.dump(data, f, indent=2)
        self.save_path = path

    @classmethod
    def load(cls, path: str) -> "Project":
        with open(path, encoding="utf-8") as f:
            data = json.load(f)

        project = cls(
            name=data["name"],
            point_count=data["point_count"],
            point_distribution=data["point_distribution"],
            border_exclusion=data["border_exclusion"],
            border_rect=data.get("border_rect"),
            border_polygon=data.get("border_polygon"),
            coral_codes=data["coral_codes"],
            coral_groups=data.get("coral_groups", []),
            species_list=data.get("species_list", []),
        )

        def _load_annotation(a_data: dict) -> ImageAnnotation:
            ann = ImageAnnotation(
                image_path=a_data["image_path"],
                image_width=a_data["image_width"],
                image_height=a_data["image_height"],
                scale_factor=a_data.get("scale_factor", 1.0),
                scale_unit=a_data.get("scale_unit", "cm"),
            )
            for p_data in a_data["points"]:
                ann.points.append(Point(**p_data))
            for m_data in a_data.get("measurements", []):
                ann.measurements.append(Measurement(
                    id=m_data["id"],
                    type=m_data["type"],
                    label=m_data["label"],
                    points=[tuple(pt) for pt in m_data["points"]],
                    value=m_data["value"],
                    unit=m_data["unit"],
                    species=m_data.get("species", ""),
                    auto_width=m_data.get("auto_width", 0.0),
                    auto_height=m_data.get("auto_height", 0.0),
                    area=m_data.get("area", 0.0),
                    perimeter_len=m_data.get("perimeter_len", 0.0),
                    angle=m_data.get("angle", 0.0),
                ))
            return ann

        if "stations" in data:
            for s_data in data["stations"]:
                station = Station(
                    name=s_data["name"],
                    depth_m=s_data.get("depth_m"),
                    date=s_data.get("date"),
                    gps_lat=s_data.get("gps_lat"),
                    gps_lon=s_data.get("gps_lon"),
                    notes=s_data.get("notes", ""),
                )
                for a_data in s_data.get("annotations", []):
                    station.annotations.append(_load_annotation(a_data))
                project.stations.append(station)
        elif "annotations" in data:
            # Old flat format — auto-migrate to a single station
            station = Station(name="Station 1")
            for a_data in data["annotations"]:
                station.annotations.append(_load_annotation(a_data))
            project.stations.append(station)

        project.save_path = path
        return project

File: rust/src/models/project.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Point {
    pub x: f64,
    pub y: f64,
    pub index: i64,
    #[serde(default)]
    pub label: Option<String>,
    #[serde(default)]
    pub category: Option<String>,
}

/// "line" | "polyline" | "polygon"
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Measurement {
    pub id: String,
    #[serde(rename = "type")]
    pub kind: String,
    /// fragment name, e.g. "Frag-01"
    pub label: String,
    /// image-space pixel coords
    pub points: Vec<(f64, f64)>,
    /// length (unit) or area (unit^2)
    pub value: f64,
    /// "cm" or "m"
    pub unit: String,
    #[serde(default)]
    pub species: String,
    /// oriented/bbox width in real units
    #[serde(default)]
    pub auto_width: f64,
    /// oriented/bbox height in real units
    #[serde(default)]
    pub auto_height: f64,
    /// polygon area in real units^2
    #[serde(default)]
    pub area: f64,
    /// polygon perimeter in real units
    #[serde(default)]
    pub perimeter_len: f64,
    /// tilt of the height axis, degrees from horizontal
    #[serde(default)]
    pub angle: f64,
}

impl Measurement {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        kind: impl Into<String>,
        label: impl Into<String>,
        points: Vec<(f64, f64)>,
        value: f64,
        unit: impl Into<String>,
        species: impl Into<String>,
        auto_width: f64,
        auto_height: f64,
        area: f64,
        perimeter_len: f64,
        angle: f64,
    ) -> Self {
        Measurement {
            id: Uuid::new_v4().to_string(),
            kind: kind.into(),
            label: label.into(),
            points,
            value,
            unit: unit.into(),
            species: species.into(),
            auto_width,
            auto_height,
            area,
            perimeter_len,
            angle,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImageAnnotation {
    pub image_path: String,
    #[serde(default)]
    pub points: Vec<Point>,
    #[serde(default)]
    pub image_width: i64,
    #[serde(default)]
    pub image_height: i64,
    /// pixels per scale_unit; 1.0 = not calibrated
    #[serde(default = "default_scale_factor")]
    pub scale_factor: f64,
    /// "cm" or "m"
    #[serde(default = "default_scale_unit")]
    pub scale_unit: String,
    #[serde(default)]
    pub measurements: Vec<Measurement>,
}

fn default_scale_factor() -> f64 {
    1.0
}

fn default_scale_unit() -> String {
    "cm".to_string()
}

impl ImageAnnotation {
    pub fn new(image_path: impl Into<String>) -> Self {
        ImageAnnotation {
            image_path: image_path.into(),
            points: Vec::new(),
            image_width: 0,
            image_height: 0,
            scale_factor: 1.0,
            scale_unit: "cm".to_string(),
            measurements: Vec::new(),
        }
    }

    pub fn labeled_count(&self) -> usize {
        self.points.iter().filter(|p| p.label.is_some()).count()
    }

    pub fn is_complete(&self) -> bool {
        !self.points.is_empty() && self.labeled_count() == self.points.len()
    }

    /// Percentage coverage per label, rounded to 2 decimal places.
    pub fn coverage_stats(&self) -> HashMap<String, f64> {
        let labeled: Vec<&Point> = self.points.iter().filter(|p| p.label.is_some()).collect();
        if labeled.is_empty() {
            return HashMap::new();
        }
        let total = labeled.len() as f64;
        let mut counts: HashMap<String, u32> = HashMap::new();
        for p in &labeled {
            let key = p.label.clone().unwrap_or_default();
            *counts.entry(key).or_insert(0) += 1;
        }
        counts
            .into_iter()
            .map(|(k, v)| (k, (v as f64 / total * 100.0 * 100.0).round() / 100.0))
            .collect()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Station {
    pub name: String,
    #[serde(default)]
    pub depth_m: Option<f64>,
    /// ISO-8601: "2024-03-15"
    #[serde(default)]
    pub date: Option<String>,
    #[serde(default)]
    pub gps_lat: Option<f64>,
    #[serde(default)]
    pub gps_lon: Option<f64>,
    #[serde(default)]
    pub notes: String,
    #[serde(default)]
    pub annotations: Vec<ImageAnnotation>,
}

impl Station {
    pub fn new(name: impl Into<String>) -> Self {
        Station {
            name: name.into(),
            depth_m: None,
            date: None,
            gps_lat: None,
            gps_lon: None,
            notes: String::new(),
            annotations: Vec::new(),
        }
    }

    pub fn total_points(&self) -> usize {
        self.annotations.iter().map(|a| a.points.len()).sum()
    }

    pub fn labeled_points(&self) -> usize {
        self.annotations.iter().map(|a| a.labeled_count()).sum()
    }

    pub fn is_complete(&self) -> bool {
        !self.annotations.is_empty() && self.annotations.iter().all(|a| a.is_complete())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CoralGroup {
    pub name: String,
    pub codes: Vec<String>,
    /// 6-character hex color (no '#'), when known from a CPCe code-file import.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub color: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Project {
    pub name: String,
    #[serde(default = "default_point_count")]
    pub point_count: i64,
    /// random | stratified | uniform
    #[serde(default = "default_distribution")]
    pub point_distribution: String,
    /// uniform pixel border to exclude
    #[serde(default)]
    pub border_exclusion: i64,
    /// [x_min, y_min, x_max, y_max] if set by click
    #[serde(default)]
    pub border_rect: Option<Vec<f64>>,
    /// [[x, y], ...] if set by polygon drawing
    #[serde(default)]
    pub border_polygon: Option<Vec<Vec<f64>>>,
    #[serde(default)]
    pub coral_codes: HashMap<String, String>,
    #[serde(default)]
    pub coral_groups: Vec<CoralGroup>,
    /// known species/genus names
    #[serde(default)]
    pub species_list: Vec<String>,
    #[serde(default)]
    pub stations: Vec<Station>,
    #[serde(skip)]
    pub save_path: Option<String>,
}

fn default_point_count() -> i64 {
    10
}

fn default_distribution() -> String {
    "random".to_string()
}

/// Old flat save format, auto-migrated into a single "Station 1" on load.
#[derive(Debug, Deserialize)]
struct LegacyProjectFile {
    annotations: Vec<ImageAnnotation>,
}

impl Project {
    pub fn new(name: impl Into<String>) -> Self {
        Project {
            name: name.into(),
            point_count: 10,
            point_distribution: "random".to_string(),
            border_exclusion: 0,
            border_rect: None,
            border_polygon: None,
            coral_codes: HashMap::new(),
            coral_groups: Vec::new(),
            species_list: Vec::new(),
            stations: Vec::new(),
            save_path: None,
        }
    }

    /// Flat view across all stations — for statistics and export.
    pub fn annotations(&self) -> Vec<&ImageAnnotation> {
        self.stations.iter().flat_map(|s| &s.annotations).collect()
    }

    pub fn save(&mut self, path: impl AsRef<Path>) -> Result<()> {
        let path = path.as_ref();
        let json = serde_json::to_string_pretty(self).context("serializing project")?;
        fs::write(path, json).with_context(|| format!("writing {}", path.display()))?;
        self.save_path = Some(path.to_string_lossy().into_owned());
        Ok(())
    }

    pub fn load(path: impl AsRef<Path>) -> Result<Self> {
        let path = path.as_ref();
        let text = fs::read_to_string(path).with_context(|| format!("reading {}", path.display()))?;
        let value: serde_json::Value =
            serde_json::from_str(&text).context("parsing project JSON")?;

        let mut project: Project = if value.get("stations").is_some() {
            serde_json::from_value(value).context("deserializing project")?
        } else {
            // Old flat format — auto-migrate to a single station.
            let legacy: LegacyProjectFile =
                serde_json::from_value(value.clone()).context("deserializing legacy project")?;
            let mut project: Project = serde_json::from_value(value).context("deserializing project header")?;
            project.stations = vec![Station {
                name: "Station 1".to_string(),
                depth_m: None,
                date: None,
                gps_lat: None,
                gps_lon: None,
                notes: String::new(),
                annotations: legacy.annotations,
            }];
            project
        };

        project.save_path = Some(path.to_string_lossy().into_owned());
        Ok(project)
    }
}
